### PR TITLE
Thicker text outlines support using multi render passes and stencil tests 

### PR DESCRIPTION
**Pull request by raytranuk**
Related to: https://github.com/playcanvas/editor/issues/12

API changes:
`pc.ElementComponent` `pc.ELEMENTTYPE_TEXT` `outlineThickness` property's max value extended from 1.0 to 5.0

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/graphics/program-lib/chunks/msdf.frag [hunk 1]
```diff
@@ -35,17 +35,24 @@ vec4 applyMsdf(vec4 color) {
...
+        //float smoothing = clamp(w.x * font_textureWidth / font_pxrange, 0.0, 0.5);
     #else
         float font_size = 16.0; // TODO fix this
         // smoothing gets smaller as the font size gets bigger
         // don't have fwidth we can approximate from font size, this doesn't account for scaling
         // so a big font scaled down will be wrong...
-
-        float smoothing = clamp(font_pxrange / font_size, 0.0, 0.5);
+        float smoothing = 0.0; // smoothing doesn't work with multi-pass stencil rendering
+        //float smoothing = clamp(font_pxrange / font_size, 0.0, 0.5);
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/msdf.frag]: Slightly concerned about this. Does setting smoothing to 0 make edges look noticeably worse? Do you start seeing nasty aliasing when the text is reasonably small?

Comment (raytranuk) [reply src/graphics/program-lib/chunks/msdf.frag]: I did a comparison with smaller text with and without smoothing and it does look more aliased when using stencils without smoothing - unacceptably so IMO - but unfortunately, smoothing doesn't work with stenciling - so we will have to think up another solution that works with smoothing/anti-aliasing.

Comment (raytranuk) [reply src/graphics/program-lib/chunks/msdf.frag]: I think I can just apply the smoothing in the alpha - I will try it

Comment (raytranuk) [reply src/graphics/program-lib/chunks/msdf.frag]: Code updated to now work with smoothing! :)

File: src/framework/components/element/text-element.js [hunk 2]
```diff
@@ -230,12 +245,91 @@ Object.assign(pc, function () {
             }
         },
 
+        _updateMeshInfos: function () {
+            var i;
+            var len;
+
+            var multiPassNeeded = (this._outlineThickness > 1.0 && this._font.textures.length == 1);
```
[the remark this answers is not in the document]
Comment (raytranuk) [reply src/framework/components/element/text-element.js:252]: This is a problem (for some reason I assumed it was just non-msdf fonts that had multiple textures) - I will have to refactor a bunch of this code...

File: src/framework/components/element/text-element.js [hunk 3]
```diff
@@ -230,12 +245,91 @@ Object.assign(pc, function () {
...
+                    // keep existing entry but set correct parameters to mesh instance
+                    var mi = this._meshInfo[i].meshInstance;
+                    if (mi) {
+                        mi.setParameter("font_sdfIntensity", this._font.intensity);
+                        mi.setParameter("font_pxrange", this._getPxRange(this._font));
+                        mi.setParameter("font_textureWidth", this._font.data.info.maps[ti].width);
+                        this._setTextureParams(mi, this._font.textures[ti]);
+
+                        if (multiPassNeeded) {
+                            mi.setParameter("render_pass", RenderPass.TEXT);
```
[the remark this answers is not in the document]
Comment (raytranuk) [reply src/framework/components/element/text-element.js:270]: This bit of code handles going from single pass to multi pass (or vice versa) without recreating all meshInstances - so when going from single to multi pass, the first pass should always be `RenderPass.TEXT`

File: src/framework/components/element/text-element.js [hunk 4]
```diff
@@ -373,43 +481,52 @@ Object.assign(pc, function () {
...
-                        meshInfo.normals[v * 4 * 3 + 10] = 0;
-                        meshInfo.normals[v * 4 * 3 + 11] = -1;
+                    var mesh = null;
+
+                    switch (this._meshInfo[i].renderPass) {
+                        case RenderPass.TEXT:
+                        case RenderPass.ALL_IN_ONE:
+                                // set up indices and normals whose values don't change when we call _updateMeshes
+                            for (var v = 0; v < l; v++) {
+                                    // create index and normal arrays since they don't change
```
[the remark this answers is not in the document]
Comment (raytranuk) [reply src/framework/components/element/text-element.js:491]: I will fix this

[the remark this answers is not in the document]
Comment (raytranuk) [reply src/framework/components/element/text-element.js]: fixed

File: src/framework/components/element/text-element.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (raytranuk) [reply src/framework/components/element/text-element.js]: refactored a bunch of code to fix this - thanks!

File: src/framework/components/element/text-element.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (raytranuk) [reply src/framework/components/element/text-element.js]: I've refactored this code - should make more sense?

**Comment by yaustar** [comment]
Could we allow for WebGL2 and limit it for WebGL1?

**Pull request closed without merging**